### PR TITLE
Clearer management of files when assigning typesetting

### LINKED ISSUE #138: Author Uploaded Proofing Files not available to Typesetters
if an author uploads their proofed article as a Proofing file (and they almost always do) then this type of file is not available for Editors to assign to typesetters for corrections. Only MS or Production Files are currently assignable at this stage.

### LINKED ISSUE #140: Add an upload button for further proofs on the typesetting assignment page
Sometimes editors will find that there is a missing proofing file when requesting corrections, which leads to them having to go back to the previous screen to upload a file. We should let them upload a new proof from the assignment screen itself.

### LINKED ISSUE #152: Allow editors to select what proofing comments/files are shared with the typesetters
When creating a correction task (new typesetting task), the editor has the option of sharing all of the proofing comments/files or None. In some cases the editor may want to share only the comments/files from one of the proofreaders but not others (e.g when the editor does a final round of proofing where they collate down all the corrections asked from the authors).

In that case, there should be a checkbox next to each proofreading comment/file controlling what will be shared with the typesetter.

### LINKED ISSUE #154: Correction Files
Editors currently must select a file when creating a correction task, this should not be required anymore.

### LINKED ISSUE #166: Improve visibility of proofed files when assigning typesetting corrections
### Problem reported

When creating round 2 typesetting assignments, editors report confusion about whether files that have been uploaded by proofreaders will be available to typesetters. The trouble arises on the assignment creation page when they are trying to manage and distinguish between three different places where files are specified:

1. Which file to typeset -- usually a manuscript file in Word format. In rounds after the first, the requirement to check such a file again causes some consternation, since it seems that by doing so one is designating a new file, when the file they want to highlight as containing new information is actually the proofed file (below).

2. The proofreader-uploaded files. These files are hidden at first, so it's not immediately clear that this file will even be included (see screenshot).

Screenshot from 2022-07-25 10-57-33

3. The previously uploaded typeset files. These are listed under "Corrections". They are automatically selected, in contrast to the files to typeset.

### Suggested changes 

To the assignment page

- [x] List any proofreader-uploaded files on the main assignment creation page--that is, the user does not have click "View proofreader feedback" to see them.
- [x] Automatically check the file to typeset that has already been sent to the typesetter. In other words, the form should remember what files were sent in the previous typesetting round, as it does for the typeset files at the bottom of the page. This is important because it helps the user understand that this file has already been sent and they can leave it alone.
- [x] While you're at it, standardize the styling of these three file lists so that they are the same (they all use different Foundation CSS elements at the moment)

To the typesetter view

- [x] Add a "Modified" column to the "Files to Typeset" list so typesetters can tell when a file is old or new.

### LINKED ISSUE #194: On assignment creation screen, list Typeset Files in the "2. Select Files [for typesetting]" field
On the screen for creating a new typesetting assignment, even if I'm creating a new typesetting round for a _revised_ PDF (or XML) I'm still forced to select a manuscript file in the "2. Select Files [for typesetting]" field. This is less than ideal, because for a 2pass PDF, the source file is the 1pass PDF, _not_ the original MS--so indicating the MS as the source file might confuse the typesetter. In addition, the MS might no longer be accurate because further revisions were likely requested during the PDF/XML proofing process. 

Therefore, I suggest removing the "5. Corrections" section and instead **expanding the "2. Select Files" section to include Typeset Files in addition to the manuscripts/copyedited files**. In my opinion, it would be more intuitive if, when requesting typesetting for a 2pass PDF, I could select the annotated-with-corrections 1pass PDF as my source file instead of a potentially out-of-date manuscript. Or, if I'm requesting 1pass XML, I might select the final/approved PDF as the source file. This would also provide more clarity/transparency for the PM creating the assignment, because the "2. Select Files" field does show full info for each of the available files, including upload/modify dates (whereas "5. Corrections" lists the file's Label and nothing else).

I suppose a simpler alternative might just be to _not_ make "2. Select Files" a required field, and to leave the "5. Corrections" section as-is. Assuming we can get Apex on board with re-versioning Typeset Files instead of uploading duplicate versions of each format, we can probably make do with just the Label displayed in section 5.

### PR DESCRIPTION
**Pull request by joemull**
This tackles a set of related issues with file management during typesetting assignments, especially in subsequent rounds of typesetting involving proofing files and galley correction.

- Closes #166 
- Closes #194 
- Closes #152 
- Closes #154
- Closes #140
- Closes #138
- Closes BirkbeckCTP/janeway#4314

Regarding #152, this PR does not allow editors to individually select proofreaders' comments, but it does allow them to individually select proofing files separately from comments, and upload their own proofing files. If they want to collate the comments together, they can do this more easily given the summary/details display of proofing comments.

Dependencies:

- BirkbeckCTP/janeway#4327
- BirkbeckCTP/janeway#4328

## Editor's main view of typesetting files

image

## Editor assigns a typesetter

image
image
image
image

## Typesetter views the assignment

image
image

### CODE REVIEW

File: logic.py [hunk 1]
```diff
@@ -22,11 +23,13 @@
 
 def production_ready_files(article, file_objects=False):
     """
+    Deprecated.
     Gathers a list of production ready files.
     :param article: an Article object
     :param file_objects: Boolean
     :return: a list of File type objects
     """
+    raise DeprecationWarning('Use get_typesetting_files instead.')
```
Comment (ajrbyers) [inline logic.py:32]: Just a note: this will mean a major version bump for the typesetting plugin.

File: views.py [hunk 2]
```diff
@@ -278,6 +278,8 @@ def typesetting_upload_galley(request, article_id, assignment_id=None):
         )
 
     if 'prod' in request.POST:
```
Comment (ajrbyers) [inline views.py:280]: This should just be removed.

**Review by ajrbyers — CHANGES_REQUESTED** [review]
Thanks for this @joemull I think this will help editors out a lot with this step in the process.

- I've added two review comments
- I do have a concern around not presenting the typesetters, automatically, with figure files. Do you think editors will be okay with selecting them? I think journals like CG will consider this a step backwards as they often have large numbers of images that will need to be selected.
- For articles in typesetting we will need a data migration that will add the figures to the list of files for the typesetter otherwise they will lose access to them.

Let me know what you think about the above.

File: logic.py (hunk 1 above)
Comment (joemull) [reply logic.py:32]: I only learned this recently but warnings don't cause an exception or terminate the program, so this is what we want to do some time *before* advancing the major version. Removing the function entirely is the change to make when we do a major version bump.

https://docs.python.org/3/library/warnings.html

Comment (ajrbyers) [reply logic.py:32]: > I only learned this recently but warnings don't cause an exception or terminate the program, so this is what we want to do some time _before_ advancing the major version. Removing the function entirely is the change to make when we do a major version bump.
> 
> https://docs.python.org/3/library/warnings.html

Ace!

**Comment by joemull** [comment]
>     * I do have a concern around not presenting the typesetters, automatically, with figure files. Do you think editors will be okay with selecting them? I think journals like CG will consider this a step backwards as they often have large numbers of images that will need to be selected.

I think we got into trouble with the usability of this interface by automatically controlling the visibility of things, so what if we make a button with which editors can select all figure files, if they want to? I think I can expand the select_all widget to support looking for value in a `data-` attribute, and we could  use the result of `file_type` there.

>     * For articles in typesetting we will need a data migration that will add the figures to the list of files for the typesetter otherwise they will lose access to them.

I can write a migration for this.

**Comment by joemull** [comment]
Or we could add a setting that controls whether they are pre-selected.

Or if we think most people will want this, we could just preselect them always, without a setting. A few other things are now preselected, so this would not be unique behavior.

Thoughts @ajrbyers ?

**Comment by joemull** [comment]
@ajrbyers I've gone ahead and pre-checked data and figure files.

**Review by ajrbyers — CHANGES_REQUESTED** [review]
Generally this is really good.

1. When assigning a correction request the editor is presented with the option to select Galley files twice. They Typesetter will have access to the Galley regardless of whether the editor selects the file here.
Screenshot 2024-08-14 at 15 35 59

2. When logged in as a Typesetter and undertaking a correction task I was able to see the same Galley listed twice.
Screenshot 2024-08-14 at 15 40 18

3. The article metadata has been moved to the foot of the Typesetter page. Would it be worth adding some sort of signpost? It could easily be missed.

**Comment by joemull** [comment]
Thanks @ajrbyers.

> 1. When assigning a correction request the editor is presented with the option to select Galley files twice.

One part of the form asks "Which files should be used?" and the other asks "Which galleys need to be corrected?". The two are different things and it is important to be literal and clear in these interfaces, according to the user reports. The selected files control different lists on the typesetter's end, with different explanations (see below).

I prevent errors with the confirmable form:

https://github.com/BirkbeckCTP/typesetting/blob/07115db9480f3919d0bef0ec9ad614be38fbcc2e/forms.py#L92-L105

> They Typesetter will have access to the Galley regardless of whether the editor selects the file here.

That's how it used to work, but I changed it because that was one of the sources of confusion. The typesetter sees this text above the first list of files:

"The following files have been selected for use or reference when generating the typeset articles."

For "have been selected" to be true, it is important that all the files, and only the files, selected by the editor appear here. Otherwise the interface can contradict the editors' instructions, which is a source of confusion that users reported.

Granted, it is true that the galley files uploaded by the typesetter will always appear in a separate list. The text above that list says:

"Any galleys (typeset files) that were selected by the editor for correction, or that you originally uploaded, will appear here."

This list needs to stay separate so that the editors' selection is a meaningful piece of communication between editor and typesetter.

> 2. When logged in as a Typesetter and undertaking a correction task I was able to see the same Galley listed twice.

Thanks for catching that. I'll fix it.

> 3. The article metadata has been moved to the foot of the Typesetter page. Would it be worth adding some sort of signpost? It could easily be missed.

I've added a skip link. Doesn't look fantastic but it will do the trick for now:

image

**Review by ajrbyers — APPROVED** [review]
🥇

**Pull request merged**